hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,179 @@
+        MobConstants.DEFAULT_MOB_COMPACTION_CHORE_PERIOD), TimeUnit.SECONDS);
+    this.master = master;
+    this.conf = master.getConfiguration();
+    this.regionBatchSize =
+        master.getConfiguration().getInt(MobConstants.MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE,
+          MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE);
+
+  }
+
+  @Override
+  protected void chore() {
+
+    boolean reported = false;
+
+    try (Connection conn = ConnectionFactory.createConnection(conf);
+         Admin admin = conn.getAdmin(); ) {
+
+      if (running) {
+        LOG.warn(getName() +" is running already, skipping this attempt.");
+        return;
+      }
+      running = true;
+      TableDescriptors htds = master.getTableDescriptors();
+      Map<String, TableDescriptor> map = htds.getAll();
+      for (TableDescriptor htd : map.values()) {
+        if (!master.getTableStateManager().isTableState(htd.getTableName(),
+          TableState.State.ENABLED)) {
+          continue;
+        }
+        for (ColumnFamilyDescriptor hcd : htd.getColumnFamilies()) {
+          if (hcd.isMobEnabled()) {
+            if (!reported) {
+              master.reportMobCompactionStart(htd.getTableName());
+              reported = true;
+            }
+            LOG.info(" Major compacting "+ htd.getTableName() + " cf=" + hcd.getNameAsString());
+            if (regionBatchSize == MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE) {
+              admin.majorCompact(htd.getTableName(), hcd.getName());
+            } else {
+              performMajorCompactionInBatches(admin, htd, hcd);
+            }
+          }
+        }
+        if (reported) {
+          master.reportMobCompactionEnd(htd.getTableName());
+          reported = false;
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to compact", e);
+    } finally {
+      running = false;
+    }
+  }
+
+  private void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i=0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for(RegionInfo ri: toCompact) {
+      startCompaction(admin, htd.getTableName(), ri);
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    while(!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri: toCompact) {
+        if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
```

Comment:
if a particular region gets stuck we'll wait here indefinitely and never get around to requesting compactions for other tables.

should make sure we at least have a WARN or something after it's taken a threshold amount of time so that an operator can intervene.

Reply:
I think we will get RPC timeout, at least, otherwise I do not know how to handle this. I expect IOException here and log it as WARN, then continue with other regions.